qcodes/dataset/measurements.py

```diff
@@ -159,23 +181,34 @@ def add_result(self,
                                      f'setpoint values for {param}:'
                                      f' {stuffweneed}.'
                                      f' Values only given for {params}.')
+        if inserting_unrolled_array and inserting_as_arrays:
+            raise RuntimeError("Trying to insert multiple data values both "
+                               "in array from and as numeric. This is not "
+                               "possible.")
+        elif inserting_as_arrays:
+            input_size = 1
 
         for index in range(input_size):
             res_dict = {}
             for partial_result in res:
-                param = str(partial_result[0])
-                value = partial_result[1]
-                # For compatibility with the old Loop, setpoints are
-                # tuples of numbers (usually tuple(np.linspace(...))
-                if hasattr(value, '__len__') and not(isinstance(value, str)):
-                    value = cast(Union[Sequence,np.ndarray], value)
-                    if isinstance(value, np.ndarray):
-                        value = np.atleast_1d(value)
-                    res_dict.update({param: value[index]})
-                else:
-                    res_dict.update({param: value})
-
-            self._results.append(res_dict)
+                param_spec = self.parameters[str(partial_result[0])]
+                if param_spec.type == 'array' and index == 0:
+                    res_dict[str(partial_result[0])] = partial_result[1]
+                elif param_spec.type != 'array':
+                    param = str(partial_result[0])
+                    value = partial_result[1]
+                    # For compatibility with the old Loop, setpoints are
+                    # tuples of numbers (usually tuple(np.linspace(...))
+                    if hasattr(value, '__len__') and not(isinstance(value,
+                                                                    str)):
+                        value = cast(Union[Sequence, np.ndarray], value)
+                        if isinstance(value, np.ndarray):
+                            value = np.atleast_1d(value).ravel()
```

Comment:
when I was profiling this, I saw indeed that after the thing that I've fixed already, `atleast_1d` takes quite some time together with the `as_float` data adapter (if i call it correctly).

Reply:
Afaik this is only there to handle the case of 0D numpy arrays (i.e. np.array(1)) which is not the same as a numpy scalar but will pass is instance check for array and take this code path. We could rewrite it to make sure that it doesn't

Reply:
omg.... and i guess writing an explicit `if` statement (as opposed to just `isinstance` check) to cover the case is less efficient, hence the `atleast_1d` without any comments around for the rationale.

Reply:
Yes atleast_1d is surprisingly slow especially already 1d and higher arrays. I will replace it with a single reshape if needed

```
a = np.array((0))
a
Out[59]: array(0)
a.shape
Out[60]: ()
%%timeit
if a.ndim == 0:
    a.reshape(1)
565 ns ± 5.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
a = np.array((0,1,3))
%%timeit
if a.ndim == 0:
    a.reshape(1)
41 ns ± 1.59 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
a = np.array((0))
%%timeit
np.atleast_1d(a)
1.15 µs ± 10.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
a = np.array((0,1,3))
%%timeit
np.atleast_1d(a)
628 ns ± 11.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
Yes I would have assumed that atleast_1d was very cheap if dealing with 1d or higher arrays but it looks like it does a lot of other checks too